cpp/src/lib/proxy_client.cc

```diff
@@ -269,5 +282,401 @@ Proxy_client::get_proxy_version() {
 
   return result;
 }
+
+void
+RDMAProxy_client :: _really_write(const char* buf, const int len){
+  int flags = 0;
+  int sent;
+  int todo = len;
+  int off = 0;
+  while(todo > 0){
+    sent = rsend(_socket, &buf[off], todo, flags);
+    if (sent < 0){
+      throw proxy_exception(sent, "really_write");
+    }
+    off += sent;
+    todo -= sent;
+  }
+}
+
+void
+RDMAProxy_client :: _really_read(char* buf, const int len){
+  int flags = 0;
+  int read = 0;
+  int todo = len;
+  int off = 0;
+  while(todo > 0){
+    read = rrecv(_socket, &buf[off], todo, flags);
+    if (read < 0){
+      throw proxy_exception(read, "really_read");
+    }
+    off += read;
+    todo -= read;
+  }
+}
+
+void RDMAProxy_client::check_status(const char *function_name) {
+  // _stream.expires_at(boost::posix_time::max_date_time);
+  if (not _status.is_ok()) {
+    ALBA_LOG(DEBUG, function_name
+             << " received rc:" << (uint32_t)_status._return_code)
+      throw proxy_exception(_status._return_code, _status._what);
+  }
+}
+
+RDMAProxy_client :: RDMAProxy_client(
+    const string &ip, const string &port,
+    const boost::asio::time_traits<boost::posix_time::ptime>::duration_type &expiry_time){
+  ALBA_LOG(INFO, "RDMAProxy_client(" << ip << ", " << port << ")");
+
+  int32_t magic{1148837403};
+  int32_t version{1};
+  _socket = rsocket(AF_INET, SOCK_STREAM, 0);
+
+  _writer = [&](const char* buffer,
+                const int len) -> void {
+    _really_write(buffer,len);
+  };
+
+  _reader = [&](char* buffer,
+                const int len) -> void {
+    _really_read(buffer,len);
+  };
+
+  if (_socket < 0){
+    throw proxy_exception(-1,"socket?");
+  }
+  struct sockaddr_in serv_addr;
+
+  serv_addr.sin_family = AF_INET;
+
+  int port_i = boost :: lexical_cast<int>(port);
+  serv_addr.sin_port = htons(port_i);
+
+  int ok = inet_pton(AF_INET, ip.c_str(), &serv_addr.sin_addr);
+  if(ok <0){
+    throw proxy_exception(errno,"ip");
+  }
+  ALBA_LOG(INFO, "connecting");
+
+  //TODO: timeouts on connect.
+  ok = rconnect(_socket, (struct sockaddr *)&serv_addr, sizeof(serv_addr));
+  if (ok < 0){
+    throw proxy_exception(errno, "connect");
+  }
+
+  _really_write((const char *)(&magic)  , sizeof(int32_t));
+  _really_write((const char *)(&version), sizeof(int32_t));
+
+}
+
+tuple<vector<string>, has_more>
+RDMAProxy_client :: list_namespaces(
```

Comment:
I considered this, but decided not to. Here's my rationale:

- I think it's wrong to try to generalize/simplify before the code is finished. In this case the timeouts still needed to be added and I was rather clueless on the effects that would bring. 
- not touching the tcp flavour of the client also means 0 chance for error on that client. 
- so we have 2 cases that look similar, with 3 cases I would certainly agree but now I'm not sure.
- lifting it in a superclass will probably make the volume of the code larger.